File: Ahorn/entities/TowerWindow.jl
module TowerWindow

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/TowerWindow" TowerWindow(x::Integer, y::Integer, lightIntensity::Number=0.6, startLit::Bool=true, view::String="Sky")

const placements = Ahorn.PlacementDict(
    "sky" => Ahorn.EntityPlacement(TowerWindow),
    "stars" => Ahorn.EntityPlacement(TowerWindow),
    "dark" => Ahorn.EntityPlacement(TowerWindow)
)

function Ahorn.selection(entity::TowerWindow)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::TowerWindow, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/towerWindow.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local towerWindow = {}

towerWindow.name = "MaggyHelper/TowerWindow"
towerWindow.depth = -50
towerWindow.fieldInformation = {
    view = {
        options = {"Sky", "Stars", "Moon", "Sunrise", "Storm", "City"},
        editable = false
    },
    lightIntensity = {
        minimumValue = 0,
        maximumValue = 1
    }
}
towerWindow.placements = {
    {
        name = "sky",
        data = {
            view = "Sky",
            lightIntensity = 0.6,
            startLit = true
        }
    },
    {
        name = "stars",
        data = {
            view = "Stars",
            lightIntensity = 0.4,
            startLit = true
        }
    },
    {
        name = "dark",
        data = {
            view = "Sky",
            lightIntensity = 0.0,
            startLit = false
        }
    }
}

function towerWindow.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/tower_window_frame/default", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    return sprite
end

function towerWindow.selection(room, entity)
    return utils.rectangle(entity.x - 24, entity.y - 64, 48, 64)
end

return towerWindow
